src/bundles/peer-locations.js

```diff
@@ -300,44 +141,67 @@ class PeerLocationResolver {
       ...opts.cache
     })
 
+    this.failedAddrs = HLRU(500)
+
+    this.queue = new PQueue({
+      concurrency: opts.concurrency,
+      autoStart: true
+    })
+
     this.geoipLookupPromises = {}
   }
 
-  async lookupWithCaches (peerId, addr, getIpfs) {
-    const ipv4Tuple = Multiaddr(addr).stringTuples().find(isNonHomeIPv4)
-    if (!ipv4Tuple) {
-      throw new Error(`Unable to resolve location for non-IPv4 address ${addr}`)
-    }
+  async findLocations (peers, getIpfs) {
+    const res = {}
 
-    var ipv4Addr = ipv4Tuple[1]
+    for (const p of peers) {
+      const peerId = p.peer.toB58String()
+      const addr = p.addr.toString()
 
-    // maybe we have it cached by ipv4 address already, check that.
-    let location = await this.geoipCache.get(ipv4Addr)
-    if (location) {
-      return location
-    }
+      const ipv4Tuple = Multiaddr(addr).stringTuples().find(isNonHomeIPv4)
+      if (!ipv4Tuple) {
+        continue
+      }
+
+      const ipv4Addr = ipv4Tuple[1]
+      if (this.failedAddrs.has(ipv4Addr)) {
+        continue
+      }
+
+      // maybe we have it cached by ipv4 address already, check that.
+      const location = await this.geoipCache.get(ipv4Addr)
+      if (location) {
+        res[peerId] = location
+        continue
+      }
 
-    // no ip address cached. are we looking it up already?
-    var locPromise = this.geoipLookupPromises[ipv4Addr]
+      // no ip address cached. are we looking it up already?
+      var locPromise = this.geoipLookupPromises[ipv4Addr]
+      if (locPromise) {
+        continue
+      }
+
+      this.queue.add(async () => {
+        return new Promise(resolve => {
+          const ipfs = getIpfs()
 
-    if (!locPromise) {
-      locPromise = new Promise((resolve, reject) => {
-        const ipfs = getIpfs()
-        geoip.lookup(ipfs, ipv4Addr, (err, data) => {
-          // save it in the caches, and remove the promise.
-          this.geoipCache.set(ipv4Addr, data)
-          delete this.geoipLookupPromises[ipv4Addr]
+          geoip.lookup(ipfs, ipv4Addr, (err, data) => {
+            delete this.geoipLookupPromises[ipv4Addr]
```

Comment:
We remove it here, but can't find a place where we add it :^)

Reply:
It's here: https://github.com/ipfs-shipyard/ipfs-webui/pull/1262/files/0e15b969fbceb9c6a344b246d0f784d8f50db96e#diff-74ea233092a3be56e73ee19c79bc2064R151 :o

Reply:
I meant we are are deleting things but we never set them.
I think we need:

```js
      this.geoipLookupPromises[ipv4Addr] = this.queue.add(async () => {
        return new Promise(resolve => {
           // ...
        })
      })
```

Reply:
oh, yeah, definitely! thanks for noticing that. Can you add that when fixing the testS?

Reply:
Yup, added in https://github.com/ipfs-shipyard/ipfs-webui/pull/1262/commits/eeccc8ae651c089343f8fa66aa3cba7d682e2bbe